cmd/searcher/search/search_regex.go

```diff
@@ -341,13 +341,12 @@ func regexSearch(ctx context.Context, rg *readerGrep, zf *store.ZipFile, fileMat
 		// Fast path for only matching file paths (or with a nil pattern, which matches all files,
 		// so is effectively matching only on file paths).
 		for _, f := range files {
-			if rg.matchPath.MatchPath(f.Name) && rg.matchString(f.Name) {
-				if len(matches) < fileMatchLimit {
-					matches = append(matches, protocol.FileMatch{Path: f.Name})
-				} else {
-					limitHit = true
-					break
-				}
+			if len(matches) >= fileMatchLimit {
+				limitHit = true
+				break
+			}
+			if match := rg.matchPath.MatchPath(f.Name) && rg.matchString(f.Name); match && !isPatternNegated || !match && isPatternNegated {
```

Comment:
```suggestion
			if match := rg.matchPath.MatchPath(f.Name) && rg.matchString(f.Name); (match && !isPatternNegated) || (!match && isPatternNegated) {
```

The lack of parens is too cool for me and makes it tricky to read :-)